### DISCUSSION TITLE
Filter posts with tags in ssg

### DISCUSSION THREAD

**Discussion by MejanH** [vercel/next.js#15559] [Help] [answered]
I created my SSG by following the docs with typescript. and I added tags in the blog posts(ex: tags: ["tech", "music"]). I want simply to implement a category page in my blog. How can I fetch posts by category or single tag?

what i want is: localhost:port/tags/tech

file tree: /pages/tags/[tag].tsx

getStaticPaths: 
```
export const getStaticPaths: GetStaticPaths = async () => {
  const postsData = await getSortedPostsData()
  const paths = postsData.map(({ tags }) => ({
    params: {
      tags: tags as string[],
    },
  }));

  console.log(paths);

  return { paths, fallback: false };
};
```
getStaticProps: 
```
export const getStaticProps: GetStaticProps = async ({ params }) => {
  const data = await getSortedPostsData();

  const allPostsData = data.filter((t) =>
    t.tags.includes(params.tags as string)
  );

  return {
    props: {
      allPostsData,
    },
  };
};

```
getSortedPostData: 
```
export function getSortedPostsData() {
  const filenames = fs.readdirSync(postsDirectory);

  const allPostsData = filenames.map((filename) => {
    const id = filename.replace(/\.md$/, "");

    const fullPath = path.join(postsDirectory, filename);
    const fullContents = fs.readFileSync(fullPath, "utf8");

    const matterResult = matter(fullContents);

    return {
      id,
      ...(matterResult.data as {
        date: string;
        title: string;
        cover: string;
        tags: string[];
      }),
    };
  });

```

**Comment by leerob** [comment, marked as answer]
In `getStaticPaths`, `params.tags` shouldn't be an array. Try this. 

```js
// pages/tags/[tag].tsx
export const getStaticPaths = async () => {
  const tags = await getAllTags()
  const paths = tags.map(({ tag }) => ({
    params: {
      tag
    },
  }));

  return { paths, fallback: false };
};
```

In `getStaticProps`, I don't think `includes` will work when you're comparing two arrays of tags. Instead, you probably want something like `array2.filter(e => !array1.includes(e))`.

**Reply by MejanH** [reply]
I can't return string instead of array in params. it still returning an array. its saying (tags) not provided as a string.

```
// /tags/[tags].tsx
export const getStaticPaths: GetStaticPaths = async () => {
  const postsData = await getSortedPostsData();

  const tags = await postsData.map(({ tags }) => {
    return tags;
  });

  console.log(tags);

  const paths = tags.map((t) => ({
    params: {
      tags: t,
    },
  }));
  console.log(paths);
  return { paths, fallback: false };
};

```
![image](https://user-images.githubusercontent.com/41269045/88680018-e1226b80-d111-11ea-9549-af3e7a1aad4b.png)

**Reply by leerob** [reply]
You have an array of arrays. You only need to create three pages here, one for each tag. So the array you pass to create the paths should only be `['tech', 'music', 'business']`.

**Reply by MejanH** [reply]
Please forgive me, I am new to Next JS. What do you mean by I have to create three pages? should I change my file tree like /pages/[tags]/[tag].tsx or something else?

**Reply by leerob** [reply]
When `getStaticPaths` runs, it will create three different pages.

1. `/pages/tags/tech`
1. `/pages/tags/music`
1. `/pages/tags/business`

For example...

```js
// pages/tags/[tag].tsx 
export const getStaticPaths = () => {
  const tags = ['tech', 'music', 'business'];

  const paths = tags.map((tag) => ({
    params: {
      tag
    },
  }));

  return { paths, fallback: false };
};
```
Note that the dynamic value is `tag`, not `tags`, based on the filename and `[tag]`. [Here's a video](https://youtu.be/wk-dUsLUADg?t=873) with an example.

**Comment by MejanH** [comment]
Thanks, it worked.

My problem was, I thought i could filter posts by an array. after I convert my multiple arrays of tags into a single array it worked.

So, filtering with multiple strings from multiple arrays is not possible with `getStaticPaths`?

my tags are like this:

[
['tech', music'],
['music', 'song'],
['business']
]
